### PR TITLE
Raise error not string

### PR DESCRIPTION
**Pull request by CaedenPH**
### Describe your change:
Replaces string errors with raising exceptions

* [ ] Add an algorithm?
* [x] Fix a bug or typo in an existing algorithm?
* [ ] Documentation change?

### Checklist:
* [x] I have read [CONTRIBUTING.md](https://github.com/TheAlgorithms/Python/blob/master/CONTRIBUTING.md).
* [x] This pull request is all my own work -- I have not plagiarized.
* [x] I know that pull requests will not be merged if they fail the automated tests.
* [x] This PR only changes one algorithm file.  To ease review, please open separate PRs for separate algorithms.
* [x] All new Python files are placed inside an existing directory.
* [x] All filenames are in all lowercase characters with no spaces or dashes.
* [x] All functions and variable names follow Python naming conventions.
* [x] All function parameters and return values are annotated with Python [type hints](https://docs.python.org/3/library/typing.html).
* [x] All functions have [doctests](https://docs.python.org/3/library/doctest.html) that pass the automated testing.
* [x] All new algorithms include at least one URL that points to Wikipedia or another similar explanation.
* [x] If this pull request resolves one or more open issues then the commit message contains `Fixes: #{$ISSUE_NO}`.

### CODE REVIEW

**Comment by CaedenPH** [comment]
Apparently I've accidentally pushed the `.flake8` change in a previous pr.
Fixing...

File: .flake8 [hunk 1]
```diff
@@ -6,3 +6,4 @@ extend-ignore =
     # Formatting style for `black`
     E203  # Whitespace before ':'
     W503  # Line break occurred before a binary operator
+    B023  # Function definition does not bind loop variable
```
Comment (cclauss) [inline .flake8:9]:
```suggestion
    B023  # Function definition does not bind loop variable

```

**Comment by CaedenPH** [comment]
@cclauss The `.flake8` change is from #7941

**Comment by CaedenPH** [comment]
Fixing checks...

**Comment by CaedenPH** [comment]
@cclauss I'll apply that `__repr__` structure to what I've changed so far

**Comment by CaedenPH** [comment]
Pre-commit failed due to error mentioned in #7940 but build workflow is :+1:

File: data_structures/binary_tree/segment_tree_other.py [hunk 2]
```diff
@@ -27,29 +27,29 @@ class SegmentTree:
     >>> for node in num_arr.traverse():
     ...     print(node)
     ...
```
Comment (cclauss) [inline data_structures/binary_tree/segment_tree_other.py:29]:
```suggestion
    >>> tuple(num_arr.traverse())  # doctest: +NORMALIZE_WHITESPACE
    (
        SegmentTreeNode(start=0, end=4, val=15),
        [ ... ]
    )
```

**Comment by CaedenPH** [comment]
@cclauss I made the requested changes - pre-commit still failing due to the `B023` error

File: genetic_algorithm/basic_string.py [hunk 3]
```diff
@@ -129,9 +129,7 @@ def select(parent_1: tuple[str, float]) -> list[str]:
             child_n = int(parent_1[1] * 100) + 1
             child_n = 10 if child_n >= 10 else child_n
             for _ in range(child_n):
-                parent_2 = population_score[  # noqa: B023
-                    random.randint(0, N_SELECTED)
-                ][0]
+                parent_2 = population_score[random.randint(0, N_SELECTED)][0]
```
Comment (cclauss) [inline genetic_algorithm/basic_string.py:132]:
```suggestion
                parent_2 = random.choice(population_score[:N_SELECTED])[0]
```

**Comment by CaedenPH** [comment]
Pre-commit is alive!

**Review by cclauss — APPROVED** [review]

**Pull request merged**